firebase_auth: Incompatible with Windows on NET 9 or VS 17.12.0

### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Which plugins are affected?

Auth

### Which platforms are affected?

Windows

### Description

I just updated to the latest Visual Studio version and this occured on Flutter 3.24.3 and then I upgraded to 3.24.4 and it was still there. Been running this on every project between tries:

# Global
dart pub cache clean --force
flutter update-packages --force-upgrade

# Once per project
flutter clean
flutter pub upgrade
flutter pub get

And then I tried adding a simple project with just the firebase_auth package and it simply did not work.

# .NET / Visual Studio Versions
dotnet --version = 9.0.100
Visual Studio Community 2022 17.12.0

Will try to rollback to previous DOTNET / VS version to see if there's any change.

### Reproducing the issue

flutter create fb_test
cd fb_test
flutter pub add firebase_auth
flutter build windows --debug

### Firebase Core version

3.7.0

### Flutter Version

3.24.4

### Relevant Log Output

```shell
Z:\fb_test\windows\flutter\ephemeral\cpp_client_wrapper\include\flutter\encodable_value.h(199,60): error C2665: 'std::variant<std::monostate,bool,int32_t,int64_t,double,std::string,std::vector<uint8_t,std::allocator<uint8_t>>,std::vector<int32_t,std::allocator<int>>,std::vector<int64_t,std::allocator<int64_t>>,std::vector<double,std::allocator<double>>,flutter::EncodableList,flutter::EncodableMap,flutter::CustomEncodableValue,std::vector<float,std::allocator<float>>>::variant': no overloaded function could convert all the argument types [Z:\fb_test\build\windows\x64\plugins\firebase_auth\firebase_auth_plugin.vcxproj]
```


### Flutter dependencies

<!--- Look below for instructions on how to share your Flutter Dependencies. --->

<details>
<summary>Expand <code>Flutter dependencies</code> snippet</summary>
<br>

```yaml

Dart SDK 3.5.4
Flutter SDK 3.24.4
fb_test 1.0.0+1

dependencies:
- firebase_auth 5.3.2 [firebase_auth_platform_interface firebase_auth_web firebase_core firebase_core_platform_interface flutter meta]
- flutter 0.0.0 [characters collection material_color_utilities meta vector_math sky_engine]

dev dependencies:
- flutter_lints 4.0.0 [lints]
- flutter_test 0.0.0 [flutter test_api matcher path fake_async clock stack_trace vector_math leak_tracker_flutter_testing async boolean_selector characters collection leak_tracker leak_tracker_testing material_color_utilities meta source_span stream_channel string_scanner term_glyph vm_service]

transitive dependencies:
- _flutterfire_internals 1.3.45 [collection firebase_core firebase_core_platform_interface flutter meta]
- async 2.12.0 [collection meta]
- boolean_selector 2.1.2 [source_span string_scanner]
- characters 1.3.0
- clock 1.1.2
- collection 1.18.0
- fake_async 1.3.2 [clock collection]
- firebase_auth_platform_interface 7.4.8 [_flutterfire_internals collection firebase_core flutter meta plugin_platform_interface]
- firebase_auth_web 5.13.3 [firebase_auth_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins http_parser meta web]
- firebase_core 3.7.0 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_core_platform_interface 5.3.0 [collection flutter flutter_test meta plugin_platform_interface]
- firebase_core_web 2.18.1 [firebase_core_platform_interface flutter flutter_web_plugins meta web]
- flutter_web_plugins 0.0.0 [flutter characters collection material_color_utilities meta vector_math]
- http_parser 4.0.2 [collection source_span string_scanner typed_data]
- leak_tracker 10.0.5 [clock collection meta path vm_service]
- leak_tracker_flutter_testing 3.0.5 [flutter leak_tracker leak_tracker_testing matcher meta]
- leak_tracker_testing 3.0.1 [leak_tracker matcher meta]
- lints 4.0.0
- matcher 0.12.16+1 [async meta stack_trace term_glyph test_api]
- material_color_utilities 0.11.1 [collection]
- meta 1.15.0
- path 1.9.1
- plugin_platform_interface 2.1.8 [meta]
- sky_engine 0.0.99
- source_span 1.10.0 [collection path term_glyph]
- stack_trace 1.12.0 [path]
- stream_channel 2.1.2 [async]
- string_scanner 1.4.0 [source_span]
- term_glyph 1.2.1
- test_api 0.7.3 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph]
- typed_data 1.4.0 [collection]
- vector_math 2.1.4
- vm_service 14.3.1
- web 1.1.0

```

</details>


### Additional context and comments

_No response_

> Hey guys,
> 
> This solved the problem for me:
> 
> 1- Uninstall Visual Studio. 2- Download the installer and execute from PowerShell this command to install the version 17.8: `.\visualstudiosetup.exe --channelUri https://aka.ms/vs/17/release.LTSC.17.8/channel`

For those getting `warning LNK4099: PDB '' was not found with ` use this [this](https://github.com/firebase/flutterfire/issues/10992#issuecomment-1642114503) solution

Can someone provide the full template instantiation context from the error message? This will be in the full build log / output window, but may not appear in a summary error log / error window.

The initially quoted error message is pointing to `Z:\fb_test\windows\flutter\ephemeral\cpp_client_wrapper\include\flutter\encodable_value.h(199,60)` as the *proximate* location of the error, which I tracked down to https://github.com/flutter/engine/blob/cea11063ae318691f0ede4ef12f2978ad01aff70/shell/platform/common/client_wrapper/include/flutter/encodable_value.h#L199 , but they're just wrapping `std::variant`. The problem appears to be within `firebase_auth`'s usage of `EncodableValue` (which in turn uses `std::variant`), and the error message mentions the top-level project file `Z:\fb_test\build\windows\x64\plugins\firebase_auth\firebase_auth_plugin.vcxproj`, but I need to see the template instantiation context in order to tell you which line of source code needs to be patched. (I don't know how to build your project; I could follow step-by-step instructions but remember that I work on C++ Standard Library code all day and I don't know the first thing about Flutter / Firebase).

"Template instantiation context" looks like this:

<details><summary>Click to expand verbose example:</summary>

```
C:\Temp>"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Auxiliary\Build\vcvarsall.bat" x64
**********************************************************************
** Visual Studio 2022 Developer Command Prompt v17.13.0-pre.1.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\Temp>type meow.cpp
#include <algorithm>
#include <vector>

struct NotComparable {};

int main() {
    std::vector<NotComparable> v;
    std::sort(v.begin(), v.end());
}

C:\Temp>cl /EHsc /nologo /W4 /c meow.cpp
meow.cpp
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8076): error C3889: call to object of class type 'std::less<void>': no matching call operator found
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\type_traits(2418): note: could be 'unknown-type std::less<void>::operator ()(_Ty1 &&,_Ty2 &&) noexcept(<expr>) const'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8076): note: Failed to specialize function template 'unknown-type std::less<void>::operator ()(_Ty1 &&,_Ty2 &&) noexcept(<expr>) const'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8076): note: With the following template arguments:
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8076): note: '_Ty1=NotComparable &'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8076): note: '_Ty2=NotComparable &'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\type_traits(2420): note: binary '<': 'NotComparable' does not define this operator or a conversion to a type acceptable to the predefined operator
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\vector(2313): note: could be 'bool std::operator <(const std::vector<_Ty,_Alloc> &,const std::vector<_Ty,_Alloc> &)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\xutility(4443): note: or       'bool std::operator <(const std::move_iterator<_Iter> &,const std::move_iterator<_Iter2> &) noexcept(<expr>)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\xutility(1905): note: or       'bool std::operator <(const std::reverse_iterator<_BidIt> &,const std::reverse_iterator<_BidIt2> &) noexcept(<expr>)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\utility(514): note: or       'bool std::operator <(const std::pair<_Ty1,_Ty2> &,const std::pair<_Uty1,_Uty2> &)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8076): note: the template instantiation context (the oldest one first) is
meow.cpp(8): note: see reference to function template instantiation 'void std::sort<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>(const _RanIt,const _RanIt)' being compiled
        with
        [
            _Ty=NotComparable,
            _RanIt=std::_Vector_iterator<std::_Vector_val<std::_Simple_types<NotComparable>>>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8236): note: see reference to function template instantiation 'void std::sort<_RanIt,std::less<void>>(const _RanIt,const _RanIt,_Pr)' being compiled
        with
        [
            _RanIt=std::_Vector_iterator<std::_Vector_val<std::_Simple_types<NotComparable>>>,
            _Pr=std::less<void>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8231): note: see reference to function template instantiation 'void std::_Sort_unchecked<NotComparable*,_Fn>(_RanIt,_RanIt,__int64,_Pr)' being compiled
        with
        [
            _Fn=std::less<void>,
            _RanIt=NotComparable *,
            _Pr=std::less<void>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8201): note: see reference to function template instantiation '_BidIt std::_Insertion_sort_unchecked<_RanIt,_Pr>(const _BidIt,const _BidIt,_Pr)' being compiled
        with
        [
            _BidIt=NotComparable *,
            _RanIt=NotComparable *,
            _Pr=std::less<void>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.43.34604\include\algorithm(8080): error C3889: call to object of class type 'std::less<void>': no matching call operator found
[...]
```
</details>

It's all of the "note" stuff after the initial "error C3889" in this case, up to the next "error C<number>" line. In my example, the STL's algorithm(8076) header is the proximate location of the error (where `std::sort()` is trying to compare elements), but the template instantiation context eventually points to `meow.cpp(8): note: see reference to function template instantiation 'void std::sort<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>(const _RanIt,const _RanIt)' being compiled` as the top-level call.

Here I want to see the top-level construction of an `EncodableValue` in `firebase_auth` since that argument type will presumably need to be changed.

--- a/packages/firebase_auth/firebase_auth/windows/firebase_auth_plugin.cpp
+++ b/packages/firebase_auth/firebase_auth/windows/firebase_auth_plugin.cpp
@@ -123,7 +123,8 @@ firebase_auth_windows::FirebaseAuthPlugin::ConvertToEncodableValue(
     case firebase::Variant::kTypeStaticBlob:
       return EncodableValue(flutter::CustomEncodableValue(variant.blob_data()));
     case firebase::Variant::kTypeMutableBlob:
-      return EncodableValue(flutter::CustomEncodableValue(variant.mutable_blob_data()));
+      return EncodableValue(
+          flutter::CustomEncodableValue(variant.mutable_blob_data()));
     default:
       return EncodableValue();
   }

--- a/packages/firebase_auth/firebase_auth/windows/firebase_auth_plugin.cpp
+++ b/packages/firebase_auth/firebase_auth/windows/firebase_auth_plugin.cpp
@@ -121,9 +121,9 @@ firebase_auth_windows::FirebaseAuthPlugin::ConvertToEncodableValue(
     case firebase::Variant::kTypeMap:
       return FirebaseAuthPlugin::ConvertToEncodableMap(variant.map());
     case firebase::Variant::kTypeStaticBlob:
-      return EncodableValue(variant.blob_data());
+      return EncodableValue(flutter::CustomEncodableValue(variant.blob_data()));
     case firebase::Variant::kTypeMutableBlob:
-      return EncodableValue(variant.mutable_blob_data());
+      return EncodableValue(flutter::CustomEncodableValue(variant.mutable_blob_data()));
     default:
       return EncodableValue();
   }

